yksmp/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+        let mut v = Vec::new();
+        for _ in 0..num {
+            let addr = self.read_u64();
+            let _stack_size = self.read_u64();
+            let record_count = self.read_u64();
+            v.push(Function { addr, record_count });
+        }
+        v
+    }
+
+    fn read_consts(&mut self, num: u32) -> Vec<u64> {
+        let mut v = Vec::new();
+        for _ in 0..num {
+            v.push(self.read_u64());
+        }
+        v
+    }
+
+    fn read_records(&mut self, num: u64, consts: &Vec<u64>) -> Vec<Record> {
+        let mut v = Vec::new();
+        for _ in 0..num {
+            let _id = self.read_u64();
+            let offset = self.read_u32();
+            self.read_u16();
+            let num_locs = self.read_u16();
+            let locs = self.read_locations(num_locs, consts);
+            // Padding
+            self.align_8();
+            self.read_u16();
+            let num_liveouts = self.read_u16();
+            let liveouts = self.read_liveouts(num_liveouts);
+            self.align_8();
+            v.push(Record {
+                offset,
+                locs,
+                liveouts,
+            });
+        }
+        v
+    }
+
+    fn read_locations(&mut self, num: u16, consts: &Vec<u64>) -> Vec<Location> {
+        let mut v = Vec::new();
+        for _ in 0..num {
+            let kind = self.read_u8();
+            self.read_u8();
+            let _size = self.read_u16();
+            let dwreg = self.read_u16();
+            self.read_u16();
+
+            let location = match kind {
+                0x01 => {
+                    self.read_i32();
+                    Location::Register(dwreg)
+                }
+                0x02 => {
+                    let offset = self.read_i32();
+                    Location::Direct(dwreg, offset)
+                }
+                0x03 => {
+                    let offset = self.read_i32();
+                    Location::Indirect(dwreg, offset)
+                }
+                0x04 => {
+                    let offset = self.read_u32();
+                    Location::Constant(offset)
+                }
+                0x05 => {
+                    let offset = self.read_i32();
+                    Location::LargeConstant(consts[usize::try_from(offset).unwrap()])
+                }
+                _ => unreachable!(),
+            };
+
+            v.push(location)
+        }
+        v
+    }
+
+    fn read_liveouts(&mut self, num: u16) -> Vec<LiveOut> {
+        let mut v = Vec::new();
+        for _ in 0..num {
+            let dwreg = self.read_u16();
+            let size = self.read_u8();
+            v.push(LiveOut { dwreg, size });
+        }
+        v
+    }
+
+    fn align_8(&mut self) {
+        self.offset += (8 - (self.offset % 8)) % 8;
+    }
+
+    fn read_u8(&mut self) -> u8 {
+        let d = u8::from_ne_bytes(self.data[self.offset..self.offset + 1].try_into().unwrap());
```

Comment:
Eventually we should make the errors bubble up and have a proper error type, but I think that will just slow us down for now.

How about add a `FIXME` for now?

Reply:
I was wondering about that. If this errors, something went seriously wrong and the program needs to crash, as I don't think we can recover from that. A guess an error would be nice, but would unlikely be useful to any user of yk.

Reply:
The interpreter could theoretically continue I suppose.

Reply:
How? If we guard fail we need the current values from the trace to continue interpretation. If we can't read the stackmap we can't get those values. And I don't want to say: `Sorry, we couldn't read the stackmap, but we'll continue interpretation with the old bogus values anyway, so if your interpreter does something completely different be warned!`.

Reply:
you are right. sorry :)